src/subtype.c

```diff
@@ -0,0 +1,569 @@
+    // to work we need to compute issub(left,right) before issub(right,left),
+    // since otherwise the issub(a, bb.ub) check in var_gt becomes vacuous.
+    bb->ub = a;  // meet(bb->ub, a)
+    return 1;
+}
+
+static int obviously_egal(jl_value_t *a, jl_value_t *b)
+{
+    if (a == b) return 1;
+    if (jl_typeof(a) != jl_typeof(b)) return 0;
+    if (jl_is_datatype(a)) {
+        jl_datatype_t *ad = (jl_datatype_t*)a, *bd = (jl_datatype_t*)b;
+        if (ad->name != bd->name) return 0;
+        size_t i, np = jl_nparams(ad);
+        if (np != jl_nparams(bd)) return 0;
+        for(i=0; i < np; i++) {
+            if (!obviously_egal(jl_tparam(ad,i), jl_tparam(bd,i)))
+                return 0;
+        }
+        return 1;
+    }
+    if (jl_is_uniontype(a)) {
+        return obviously_egal(((jl_uniontype_t*)a)->a, ((jl_uniontype_t*)b)->a) &&
+            obviously_egal(((jl_uniontype_t*)a)->b, ((jl_uniontype_t*)b)->b);
+    }
+    return !jl_is_type(a) && jl_egal(a,b);
+}
+
+static jl_value_t *simple_join(jl_value_t *a, jl_value_t *b)
+{
+    if (a == jl_bottom_type || b == (jl_value_t*)jl_any_type || obviously_egal(a,b))
+        return b;
+    if (b == jl_bottom_type || a == (jl_value_t*)jl_any_type)
+        return a;
+    if (!jl_is_type(a) || !jl_is_type(b))
+        return (jl_value_t*)jl_any_type;
+    return jl_new_struct(jl_uniontype_type, a, b);
+}
+
+static int var_gt(jl_tvar_t *b, jl_value_t *a, jl_stenv_t *e)
+{
+    jl_varbinding_t *bb = lookup(e, b);
+    if (bb == NULL)
+        return subtype_ufirst(a, b->lb, e);
+    record_var_occurrence(bb, e);
+    if (!bb->right)  // check ∀b . b>:a
+        return subtype_ufirst(a, bb->lb, e);
+    if (!((bb->ub == (jl_value_t*)jl_any_type && !jl_is_type(a)) || subtype_ufirst(a, bb->ub, e)))
+        return 0;
+    bb->lb = simple_join(bb->lb, a);
+    return 1;
+}
+
+static jl_unionall_t *rename_unionall(jl_unionall_t *u)
+{
+    jl_tvar_t *v = jl_new_typevar(u->var->name, u->var->lb, u->var->ub);
+    jl_value_t *t = NULL;
+    JL_GC_PUSH2(&v, &t);
+    t = jl_instantiate_unionall(u, (jl_value_t*)v);
+    t = jl_new_struct(jl_unionall_type, v, t);
+    JL_GC_POP();
+    return (jl_unionall_t*)t;
+}
+
+static int is_leaf_bound(jl_value_t *v)
+{
+    if (v == jl_bottom_type) return 1;
+    if (jl_is_datatype(v)) {
+        if (((jl_datatype_t*)v)->isleaftype) return 1;
+        if (((jl_datatype_t*)v)->abstract) {
+            if (jl_is_type_type(v))
+                return 1;//!jl_has_free_typevars(jl_tparam0(v));
+            return 0;
+        }
+        jl_svec_t *t = ((jl_datatype_t*)v)->parameters;
+        size_t l = jl_svec_len(t);
+        if (((jl_datatype_t*)v)->name == jl_tuple_typename) {
+            for(int i=0; i < l; i++) {
+                if (!is_leaf_bound(jl_svecref(t,i)))
+                    return 0;
+            }
+        }
+        return 1;
+    }
+    return 0;
+}
+
+static int subtype_unionall(jl_value_t *t, jl_unionall_t *u, jl_stenv_t *e, int8_t R)
```

Comment:
The correctness, especially for something this complex, is fantastic and obviously the big story here. And comments can wait until this is more finalized. But correctness is **not** useful as a criterion for deciding whether comments are useful; that's a dangerously-high bar to try to meet. A better criterion is "will we run this code for at least a few months?" Obviously we may not be there with this code yet, which is why it's fine to wait until you're more certain.

But I'm holding you to the fact that there will be comments before merger :smile:.